### PR TITLE
feat(request-transformer) change HTTP method of upstream request

### LINKED ISSUE #1631: Request Transformer should also include changing HTTP method
The Request Transformer plugin should also support changing the HTTP method when proxying the request, ie: Kong receives a `GET` request, but the same request is being proxied to the upstream with a `POST` method.

### PR DESCRIPTION
**Pull request by thefosk**
* Request-Transfomer: can change the HTTP method of the upstream request.

### Issues resolved

Fix #1631.

### CODE REVIEW

File: kong/plugins/request-transformer/access.lua [hunk 1]
```diff
@@ -303,7 +303,33 @@ local function transform_body(conf)
...
 end
 
+local function transform_method(conf)
+  if conf.http_method then
+    ngx.req.set_method(ngx["HTTP_"..conf.http_method:upper()])
+    if conf.http_method == "GET" then
+      local content_type_value = req_get_headers()[CONTENT_TYPE]
+      local content_type = get_content_type(content_type_value)
+      if content_type == ENCODED then
+        -- Also put the body into querystring
```
Comment (thibaultcha) [inline kong/plugins/request-transformer/access.lua]: Such behavior should be configurable.

File: kong/plugins/request-transformer/access.lua [hunk 2]
```diff
@@ -303,7 +303,33 @@ local function transform_body(conf)
   end
 end
 
+local function transform_method(conf)
+  if conf.http_method then
+    ngx.req.set_method(ngx["HTTP_"..conf.http_method:upper()])
```
Comment (thibaultcha) [inline kong/plugins/request-transformer/access.lua]: All `ngx` methods are cached except this one. It should be too for consistency and efficiency.

File: kong/plugins/request-transformer/schema.lua [hunk 3]
```diff
@@ -10,8 +10,18 @@ local function check_for_value(value)
   return true
 end
 
+local function check_method(value)
+  if not value then return true end
+  local method = value:upper()
+  local ngx_method = ngx["HTTP_"..method]
+  if not ngx_method then
+    return false, method.." is not supported"
```
Comment (thibaultcha) [inline kong/plugins/request-transformer/schema.lua]: By convention our codebase returns `nil` on error. Some parts are not yet updated, but new code should comply.

File: kong/plugins/request-transformer/access.lua [hunk 4]
```diff
@@ -303,7 +303,33 @@ local function transform_body(conf)
...
+    if conf.http_method == "GET" then
```
Comment (thibaultcha) [inline kong/plugins/request-transformer/access.lua]: Maybe `HEAD` too? Should probably be configurable as well.

File: kong/plugins/request-transformer/schema.lua (hunk 3 above)
Comment (thefosk) [inline kong/plugins/request-transformer/schema.lua]: Think of this value as being stored into a `is_valid` property, that's why it's `true` or `false`.

```lua
local function check_method(value)
  if not value then return true end
  local method = value:upper()
  local ngx_method = ngx["HTTP_"..method]
  if not ngx_method then
    return false, method.." is not supported"
  end
  return true
end

local is_valid, err = check_method("hello")
```

Makes sense that the return value is `boolean`, with an optional error message into `err`.

Comment (thibaultcha) [inline kong/plugins/request-transformer/schema.lua]: But that's not how this method is being used. It's passed to `func`, and it is a common Lua idiom when the second parameter is an error to return `nil` as the first one:
```lua
local ok, err = check_method("hello")
if not ok then

end
```

Returning `false` would make sense if, say, there is no error as the 2nd return value.

**Pull request merged**